### DISCUSSION TITLE
YOLOv8x: How to Reduce False Positives & Improve Generalization in Table Tennis Ball Detection?

### DISCUSSION THREAD

**Discussion by VAIBHAVAGARWAL12** [ultralytics/ultralytics#21330] [Q&A]
Hi Ultralytics team,

I'm working on a table tennis project and training a YOLOv8x model for ball detection with the goal of eventually building a score counter. However, I’m facing major issues with false positives and generalization.

**Dataset Details:**

1. 5,000 custom annotated images of table tennis balls
2. Includes a variety of lighting conditions, angles, environments, and shot types
3. Images were resized to 640×640 during dataset creation (e.g., Roboflow or similar)
4. Dataset is diverse and covers many typical game scenarios

**Training Configs:**

- Model: YOLOv8x
- Training image size: 1280
- Trained using the Ultralytics standard training pipeline

**Observations:**

1. On one specific type of video (with consistent angle and lighting), the model detects the ball quite well — but still shows false positives (detecting hands, shoes, table edges as balls).
2. On other video types (with different angles, more motion blur, etc.), the model mostly produces false positives only, and often misses the actual ball completely.
3. Interestingly, I had previously trained the same dataset using 640 image size (with YOLOv8x), and while it performed poorly on that one consistent video, it performed better on the other varied videos.
4. This inconsistency makes it hard to build a reliable detection system for scoring. My main goal is to reduce false detections and get stable, generalizable performance across all common table tennis match conditions.

**Questions:**

1. How can I improve generalization across videos with different angles and lighting?
2. How can I reduce false positives, especially for objects like shoes, hands, and net corners?

Any recommendations on whether I should go back to training at 640 vs stick with 1280?

Here’s the [dataset link](https://app.roboflow.com/vaibhav-rcmge/balls-main-dupli/3) in case it’s helpful.

Thanks so much for your work on Ultralytics and YOLO — any tips or suggestions would be really appreciated!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @VAIBHAVAGARWAL12, thank you for sharing these detailed insights and for your interest in Ultralytics 🚀! For new users and best practices, we always recommend checking out our [Docs](https://docs.ultralytics.com/), which offer a wealth of [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples, as well as answers to many frequently asked questions.

Since this appears to be a custom training ❓ question, please make sure to include as much relevant information as possible—such as sample images, training logs, and a clear description of the problem (which you've done—thanks!). If you have any training logs or specific instances where false positives occur, sharing those will help our team provide more targeted advice.

If you believe you may have encountered a 🐛 bug, we kindly ask you to provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug the issue more efficiently.

Feel free to join the Ultralytics community for real-time support and knowledge sharing:
- [Discord](https://discord.com/invite/ultralytics) 🎧 for live chat
- [Discourse](https://community.ultralytics.com/) for in-depth discussions
- [Subreddit](https://www.reddit.com/r/Ultralytics/) for sharing and learning with others

## Upgrade

Be sure to upgrade to the latest `ultralytics` package and all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to ensure your issue or question is based on the latest release:

```bash
pip install -U ultralytics
```

## Environments

You can run YOLO in any of these up-to-date, verified environments (with dependencies like [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

---

This is an automated response 🤖. An Ultralytics engineer will also review your question and provide further assistance soon!

**Reply by glenn-jocher** [reply]
Great question @VAIBHAVAGARWAL12! For table tennis ball detection with high false positive rates, here are some targeted solutions:

**For False Positives:**
- Add hard negative examples to your dataset - specifically include cropped images of hands, shoes, and table edges labeled as background
- Lower your confidence threshold during inference (try 0.3-0.5) and rely more on NMS to filter multiple detections
- Consider using a smaller model like YOLOv8s or YOLOv8m which can be less prone to overfitting on limited data

**For Generalization:**
- Implement stronger data augmentation during training - increase mosaic probability, add more geometric transforms, and vary brightness/contrast more aggressively
- Your dataset size (5k images) might be limiting - consider collecting more diverse scenarios or using synthetic data generation

**Image Size Decision:**
Since table tennis balls are small objects, training at 1280 should theoretically help, but if 640 performed better on varied videos, it suggests your model is overfitting to high-resolution features. Try training at 832 as a compromise, or stick with 640 if your target deployment videos have lower resolution.

**Quick Test:**
Run inference with `conf=0.25` and `iou=0.7` to see if adjusting these thresholds helps balance detection vs false positives before retraining.

Would you be able to share a few example images where false positives occur most frequently? This would help identify if there are specific visual patterns causing the confusion.

**Reply by VAIBHAVAGARWAL12** [reply]
Apologies for replying so late,
Thanks for your suggestions earlier. I’ve shared multiple images with you as requested — all of them demonstrate **false negatives**. Below is a breakdown for clarity:  

### 1️⃣ These screenshots are from one video (ball detection works decently but false detections aswell):  
1. <img width="1359" height="757" alt="banner" src="https://github.com/user-attachments/assets/a683e2e7-c961-4b49-95e6-0acb2f80ed86" />
2. <img width="112" height="204" alt="elbow" src="https://github.com/user-attachments/assets/b2357796-8bd5-4820-8534-a650e2978677" />
3. <img width="1345" height="746" alt="hand" src="https://github.com/user-attachments/assets/e0941191-7c50-4b3d-b106-13ad42c6d994" />
4.  <img width="345" height="569" alt="hand_2" src="https://github.com/user-attachments/assets/6a4114a1-cae9-4ed2-af60-71299f93d6fb" />
5. <img width="1295" height="711" alt="shoes" src="https://github.com/user-attachments/assets/544d95e0-32aa-4e89-a627-047ae573362c" />
6.  <img width="1293" height="719" alt="whole_hand" src="https://github.com/user-attachments/assets/5ef84180-df88-4f42-a0e0-b8c9bcf5bb09" />
7.  <img width="1298" height="698" alt="whole_hand_4-5 times" src="https://github.com/user-attachments/assets/32cc638b-9d76-4527-a402-a617ff1eacba" />
- These images show even large objects like an entire hand being mistakenly detected as the **ball**, despite being obviously too big for a ball.  Maybe because when ball is blurry in an image is seems very big in size often 2 to 3 times bigger.
- Ball detection is otherwise the best on this video.

### 2️⃣ Scoreboard issue:  
1. <img width="1352" height="755" alt="score_board" src="https://github.com/user-attachments/assets/f8b56156-d6b8-4457-8d32-3da3bbeab5e4" />
- In a similar angle but better lighting, the scoreboard **"0"** keeps getting falsely detected as the **ball** throughout the video, while real ball detections are rare.  

### 3️⃣ Different angle / different conditions:  
1. <img width="416" height="728" alt="ball,fake_ball" src="https://github.com/user-attachments/assets/c2ef9a1f-d632-45fe-adfe-94d0a49b192b" />
2.  <img width="418" height="733" alt="fake_ball" src="https://github.com/user-attachments/assets/3c228f16-6d84-4f90-97fe-b78f6ecd909c" />
3.  <img width="423" height="731" alt="net, ball, fake ball" src="https://github.com/user-attachments/assets/dc430fd7-67f5-4fd2-8146-f21df27dca52" />
- Very few actual ball detections happen here. Different angle.
- Constantly keeps detecting something as ball which actually is not a ball. Throughout the video in upper right section.
-  Below is the screenshot of what it actually is before using the model.
<img width="437" height="727" alt="no yolo" src="https://github.com/user-attachments/assets/49afa604-7471-48e9-8f29-005e4f6e148d" />

---

## ❓ Questions:  

### 1. Regarding your suggestion to add hard negative examples (cropped hands, shoes, edges labeled as background):  
Could you clarify what you mean by **"cropping"?**  
- If I crop out a hand, that cropped image will often have the hand occupy **90% or more of the pixels** (for example, a 100px image where the hand fills nearly the whole frame). Does this still help?  
- Or did you mean cropping in the sense of **masking out everything except the hand** (i.e., turning the background to black or some solid color)?  
- I ask because in a past experiment, when I cropped out only the table (trying to detect balls just on the table), my detection performance actually became worse — **no detections on the table at all**.

### 2. Regarding dataset size:  
You mentioned **5k images might be too few**. Could you recommend a rough target for the number of images I should aim for in the dataset for this kind of **small-object detection task (table tennis ball)?**  

### 3. Additional Question (Regarding Camera Angle & Lighting Conditions):

We may have the flexibility to use our own camera setup, but since matches will be conducted under different lighting conditions, I wanted to ask:  

**Would training the model specifically on the fixed camera angle where the matches will be recorded help in any way?**  

That said, I can't say with complete certainty that we will always have this freedom to fix the camera. Please consider this as just another aspect to think about.

### 4. Should we switch to newer models for ball detections like yolov11? Will it give better result?

## Thanks a lot for helping me out earlier.

**Reply by glenn-jocher** [reply]
Thanks for the detailed examples @VAIBHAVAGARWAL12! These images really help illustrate the issues.

**1. Hard Negative Clarification:**
You're right to question this - I should have been clearer. Instead of cropping objects to fill 90% of the frame, include full scenes containing hands/shoes/scoreboards but **don't label them as balls**. This teaches the model these objects are background. Your table-only cropping experiment failed because you removed important context the model needs.

**2. Dataset Size:**
For small object detection like table tennis balls, aim for **15-20k images minimum**. Your 5k is definitely limiting performance. Consider data augmentation techniques like copy-paste augmentation to artificially increase diversity.

**3. Fixed Camera Angle:**
If you can guarantee the camera setup, training specifically on that angle would significantly improve performance for that scenario. However, collect at least 70% of your data from the target angle and 30% from varied angles to maintain some generalization capability.

**4. YOLO11 Upgrade:**
Absolutely switch to YOLO11 - it offers better performance than YOLOv8, especially for small object detection. The architecture improvements in YOLO11 should help with your false positive issues.

**Immediate Actions:**
- Upgrade to YOLO11 and retrain
- Lower your confidence threshold to 0.15-0.2 for better ball detection
- Add more training data, especially scenarios similar to your problematic videos
- Consider using smaller input size (640px) initially to avoid overfitting

Would you like specific guidance on implementing any of these suggestions?
